wordpress-mobile/WordPress-Android
# Fix custom app language selection issues across various android versions

Fixes #7267, #7265 

Updating the language of the App in real time is handled differently across different versions of Android. This PR centralizes language updates to a few new classes:
- BaseActivity
- BaseListActivity
- LocaleManager

Any Activity that should be sensitive to Locale changes should extend either `BaseActivity` or `BaseListActivity`.  

`LocaleManager` contains a set of helper methods/utilities for properly persisting, fetching and applying locale settings appropriate to the active android version. 

`AppSettingsFragment` now calls `System.exit(0)` to force an instance restart of the app so the language of the app is updated instantly, _and_ to give a consistent experience across all versions of Android. 

Once localization was updated to be carried with the active context, I ran into an issue where devices running API 24 & 25 would suddenly crash with a fatal "division by zero" exception while attempting to parse and format a string containing locale-specific grouping separators. For example, the following string: `%,d people like this` would cause a fatal crash while attempting to format it using `String.format(baseString, val)`. Turns out this is a [known Java bug for Android 7.x]("https://bugs.openjdk.java.net/browse/JDK-8167567"). 

The only way around this bug is to pass in the locale to String.format(), *BUT*, Locale.getDefault() will not work when the `language` property of the Locale object contains any regional information. Specifically, it will only work if `language=en` and will not work if `language=en_US`.  Created `LocaleManager.getSafeLocale(context)` to obtain a properly formatted locale object for the workaround and updated any places where this may be a problem by searching for `%,d` in `strings.xml`.

Fix tested on emulators running the following versions: 17, 19, 22, 23, 24, 25, 26, 27

# Screenshots
## API 19
![1_screenshot_1520393588](https://user-images.githubusercontent.com/5810477/37075591-cba634be-218f-11e8-9fe6-c0ddd86c2399.png) ![1_screenshot_1520393437](https://user-images.githubusercontent.com/5810477/37075592-cd62a170-218f-11e8-8677-ad186bdba0da.png)

## API 26
![1_screenshot_1520394446](https://user-images.githubusercontent.com/5810477/37075635-04367da2-2190-11e8-9a81-588ceff5d4c8.png) ![1_screenshot_1520394454](https://user-images.githubusercontent.com/5810477/37075640-07599c94-2190-11e8-82a5-f5412431ea7c.png) ![1_screenshot_1520394451](https://user-images.githubusercontent.com/5810477/37075643-08b1ff64-2190-11e8-93d4-cbd195f03a95.png) ![1_screenshot_1520394458](https://user-images.githubusercontent.com/5810477/37075644-0a7ab9da-2190-11e8-8737-77a59f8758c1.png) ![1_screenshot_1520394465](https://user-images.githubusercontent.com/5810477/37075649-0c7c6224-2190-11e8-93e0-1565839585c6.png)

# To test:
1. Change language to Hebrew in App Settings. 
2. On devices running 17+, verify RTL layout change.
3. Verify the language is updated througout the app. Pay special attention to: 
		- Labels are updated in the filter bar of the notification tab.
		- Tag options dropdown are updated in the Reader.  
4. Change device orientation. 
5. Verify the language change persists.
6. Kill app and reopen. 
7. Verify the language change persists. 
8. Change the language back to English.
9. repeat steps 2 - 4.
